theano/tensor/tests/test_opt.py

```diff
@@ -6173,6 +6173,27 @@ def test_local_zero_div():
             assert theano.tensor.get_scalar_constant_value(output) == 0
 
 
+def test_local_sumsqr2dot():
+    G = matrix('G')
+    W = matrix('W')
+
+    y = T.sqr( W.dimshuffle('x',0,1) * G.dimshuffle(0,'x',1) ).sum(axis=(1,2))
+    MODE = theano.compile.get_default_mode().including('local_sumsqr2dot')
+
+    f = function([W, G], y, mode=MODE)
+
+    w_val = numpy.random.rand(4, 3).astype(config.floatX)
+    g_val = numpy.random.rand(5, 3).astype(config.floatX)
+
+    f_val = f(w_val, g_val)
+    f_test = function([W,G], T.dot(T.sqr(G), T.sqr(W).sum(axis=0)), mode=MODE)
+
+    assert numpy.allclose(f_val, f_test(w_val, g_val))
```

Comment:
As this test exactly what the optimization do, if the optimization do a bad conversion, we won't detect it. So for the "golden" answer, compute it like the original graph, but in numpy and not Theano.